internal/uploader/auth_utils.go

```diff
@@ -1,73 +1,128 @@
 package uploader
 
 import (
+	"errors"
+	"fmt"
 	"strings"
 
 	"github.com/openshift/assisted-service/internal/cluster/validations"
 	"github.com/openshift/assisted-service/pkg/k8sclient"
-	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 )
 
 const (
-	openshiftTokenKey = "cloud.openshift.com"
+	openshiftNamespace  = "openshift-config"
+	openshiftSecretName = "pull-secret"
+	openshiftTokenKey   = "cloud.openshift.com"
 )
 
-func getPullSecret(clusterPullSecret string, k8sclient k8sclient.K8SClient, key string) (*validations.PullSecretCreds, error) {
-	data, exists := getOCMPullSecret(k8sclient)
-	if !exists {
-		data = clusterPullSecret
+type PullSecretCreds struct {
+	ManagementCreds *validations.PullSecretCreds
+	WorkloadCreds   *validations.PullSecretCreds
+}
+
+func getPullSecretCreds(clusterPullSecret string, k8sclient k8sclient.K8SClient) (PullSecretCreds, error) {
+	ret := PullSecretCreds{}
+
+	managementCreds, managementErr := getManagementCreds(k8sclient)
+	if managementErr == nil { // Don't block if something failed, the other creds might be usable
+		ret.ManagementCreds = managementCreds
 	}
-	if data == "" {
-		return nil, errors.Errorf("failed to find pull secret for %s", key)
+
+	workloadCreds, workloadErr := getCloudOpenshiftCreds(clusterPullSecret)
+	if workloadErr == nil { // Don't block if something failed, the other creds might be usable
+		ret.WorkloadCreds = workloadCreds
+	}
+
+	if managementErr != nil && workloadErr != nil {
+		return ret, errors.Join(
+			fmt.Errorf("failed to get management creds: %w", managementErr),
+			fmt.Errorf("failed to get workload creds: %w", workloadErr),
+		)
 	}
-	pullSecret, err := validations.ParsePullSecret(data)
+
+	return ret, nil
+}
+
+// Checks if cloud.openshift.com is present in the OCM pull secret to see if the user
+// is still opted-in to sending data
+func isOCMPullSecretOptIn(k8sclient k8sclient.K8SClient) (bool, error) {
+	creds, err := getManagementCreds(k8sclient)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse pull secret")
+		if apierrors.IsNotFound(err) {
+			// Indicates the pull secret doesn't exist, so we'll assume they're opted-in
+			return true, nil
```

Comment:
I believe this is the opposite: if the pull secret is not found, then we shouldn't assume they're opted-in. It's actually how they can opt-out (by removing these credentials or not even having this pull secret [1])

[1] https://docs.openshift.com/container-platform/4.15/support/remote_health_monitoring/opting-out-of-remote-health-reporting.html#insights-operator-new-pull-secret_opting-out-remote-health-reporting

Reply:
I'm totally fine changing this, but let's double check I haven't changed that logic and it seemed to be there on purpose https://github.com/openshift/assisted-service/blob/master/internal/uploader/auth_utils.go#L71

Reply:
To clarify:

opt-in (accept that we will handle user data):
* implicit: presence of pull-secret
* explicit: DATA_COLLECTION env var ENABLED

opt-out (disagree sending us user data):
* explicit: DATA_COLLECTION env var DISABLED
* implicit: absence of pull-secret key "cloud.openshift.com" (by extension absence of pull secret too)


as long as we respect this logic, we should be fine

Reply:
As for ENABLE_DATA_COLLECTION I don't think we should consider "unset", the env var will get a default value (we can debate over the default value if necessary)

If ENABLE_DATA_COLLECTION is true, we should try to do our best to send the data (even with spoke cluster creds IMO)., but if we have no pull secret nor key we wouldn't be able to send it anyway (can throw a warning but the bottom line is that we cannot even attempt sending the data)

Reply:
In this case, presence/absence of pull secret doesn't matter when we check if events uploading is enabled (while starting the service). It just matters when we tried to upload.

I'm changing the logic in https://github.com/openshift/assisted-service/blob/master/internal/uploader/events_uploader.go#L65 and removing this method that is not called anymore https://github.com/openshift/assisted-service/blob/master/internal/uploader/auth_utils.go#L62